compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs

```diff
@@ -191,54 +210,73 @@ runRepl opts mainDar replClient ideState = do
-        -> DynFlags
+    handleStmt
+        :: DynFlags
         -> String
-        -> Int
-        -> IO (Either Error (LPat GhcPs, Type))
-    handleLine moduleNames binds dflags l i = runExceptT $ do
-        stmt <- case parseStatement l dflags of
-            POk _ lStmt -> pure (unLoc lStmt)
-            PFailed _ _ errMsg -> throwError (ParseError errMsg)
-        (bind, expr) <- maybe (throwError (UnsupportedStatement l)) pure (splitStmt stmt)
-        liftIO $ writeFileUTF8 (fromNormalizedFilePath $ lineFilePath i)
-            (renderModule dflags moduleNames i binds bind expr)
+        -> Stmt GhcPs (LHsExpr GhcPs)
+        -> ExceptT Error ReplM ()
+    handleStmt dflags line stmt = do
+        ReplState {imports, bindings, lineNumber} <- State.get
+        (bind, expr) <- maybe (throwError (UnsupportedStatement line)) pure (splitStmt stmt)
+        liftIO $ writeFileUTF8 (fromNormalizedFilePath $ lineFilePath lineNumber)
+            (renderModule dflags imports lineNumber bindings bind expr)
         -- Useful for debugging, probably best to put it behind a --debug flag
-        -- rendered <- liftIO  $readFileUTF8 (fromNormalizedFilePath $ lineFilePath i)
+        -- rendered <- liftIO $ readFileUTF8 (fromNormalizedFilePath $ lineFilePath lineNumber)
         -- liftIO $ for_ (lines rendered) $ \line ->
         --      hPutStrLn stderr ("> " <> line)
         (lfMod, tmrModule -> tcMod) <-
             maybe (throwError TypeError) pure =<< liftIO (runAction ideState $ runMaybeT $
-            (,) <$> useE GenerateDalf (lineFilePath i)
-                <*> useE TypeCheck (lineFilePath i))
+            (,) <$> useE GenerateDalf (lineFilePath lineNumber)
+                <*> useE TypeCheck (lineFilePath lineNumber))
         -- Type of the statement so we can give it a type annotation
         -- and avoid incurring a typeclass constraint.
         stmtTy <- maybe (throwError TypeError) pure (exprTy $ tm_typechecked_source tcMod)
-        scriptRes <- liftIO $ ReplClient.runScript replClient (optDamlLfVersion opts) lfMod
-        case scriptRes of
-            Right _ -> pure (bind, stmtTy)
-            Left err -> throwError (ScriptError err)
+        -- If we get an error we don’t increment lineNumber and we
+        -- do not get a new binding
+        withExceptT ScriptError $ ExceptT $ liftIO $
+            ReplClient.runScript replClient (optDamlLfVersion opts) lfMod
+        let boundVars = mkOccSet (map occName (collectPatBinders bind))
+        State.put $! ReplState
+          { imports = imports
+          , bindings = map (first (shadowPat boundVars)) bindings <> [(toTuplePat bind, stmtTy)]
+          , lineNumber = lineNumber + 1
+          }
+    handleImport
+        :: DynFlags
+        -> ImportDecl GhcPs
+        -> ExceptT Error ReplM ()
+    handleImport dflags imp = do
+        ReplState {imports, lineNumber} <- State.get
+        -- TODO[AH] Deduplicate imports.
+        let newImports = imp : imports
```

Comment:
Hm, this seems surprisingly tricky with qualified import, explicit imports, …. I wonder if we can biggypack on `-ddump-minimal-imports` (no idea how and that drops too much).

Reply:
Yeah, I haven't looked at all the options, yet. Duplicate imports don't make it fail, but only add bloat, so this didn't seem too critical.

Hmm, AFAIK `-ddump-minimal-imports` takes into account which symbols you actually use in the module. In the REPL use case that's almost always going to lead to too few imports. But, maybe we can re-use some piece of its implementation.

Another very dumb, but simple, solution could be to deduplicate on how the import declarations render. That doesn't yield a minimal set, but at least eliminates obvious duplicates.